server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/NodeShutdownAllocationDeciderTests.java

```diff
@@ -157,43 +151,103 @@ public void testCanAutoExpandToNodeThatIsNotShuttingDown() {
-    }
+    public void testAutoExpandDuringNodeReplacement() {
 
-    private ClusterState prepareState(ClusterState initialState, SingleNodeShutdownMetadata.Type shutdownType, String nodeId) {
-        final String targetNodeName = shutdownType == SingleNodeShutdownMetadata.Type.REPLACE ? randomAlphaOfLengthBetween(10, 20) : null;
-        final SingleNodeShutdownMetadata nodeShutdownMetadata = SingleNodeShutdownMetadata.builder()
-            .setNodeId(nodeId)
-            .setType(shutdownType)
-            .setReason(this.getTestName())
-            .setStartedAtMillis(1L)
-            .setTargetNodeName(targetNodeName)
-            .setGracePeriod(
-                shutdownType == SingleNodeShutdownMetadata.Type.SIGTERM
-                    ? TimeValue.parseTimeValue(randomTimeValue(), this.getTestName())
-                    : null
-            )
+        var state = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(DiscoveryNodes.builder().add(DATA_NODE).build())
+            .metadata(Metadata.builder().put(IndexMetadata.builder(indexMetadata)))
             .build();
-        NodesShutdownMetadata nodesShutdownMetadata = new NodesShutdownMetadata(new HashMap<>()).putSingleNodeMetadata(
-            nodeShutdownMetadata
+        ;
+
+        // should auto-expand when no shutdown
+        assertThatDecision(
+            decider.shouldAutoExpandToNode(indexMetadata, DATA_NODE, createRoutingAllocation(state)),
+            Decision.Type.YES,
+            "no nodes are shutting down"
+        );
+
+        // should auto-expand when shutdown/replacement is registered and no node replacement yet
+        var shutdown = createNodesShutdownMetadata(SingleNodeShutdownMetadata.Type.REPLACE, DATA_NODE.getId());
+        state = ClusterState.builder(state)
+            .metadata(Metadata.builder(state.metadata()).putCustom(NodesShutdownMetadata.TYPE, shutdown).build())
+            .build();
+        assertThatDecision(
+            decider.shouldAutoExpandToNode(indexMetadata, DATA_NODE, createRoutingAllocation(state)),
+            Decision.Type.YES,
+            "node [" + DATA_NODE.getId() + "] is preparing to be removed from the cluster, but replacement is not yet present"
+        );
+
+        // should not auto-expand when only to replacement node when joined
+        var replacementName = shutdown.get(DATA_NODE.getId()).getTargetNodeName();
+        var replacementNode = newNode(replacementName, "node-data-1", Set.of(DiscoveryNodeRole.DATA_ROLE));
+        state = ClusterState.builder(state).nodes(DiscoveryNodes.builder(state.getNodes()).add(replacementNode).build()).build();
+
+        assertThatDecision(
+            decider.shouldAutoExpandToNode(indexMetadata, DATA_NODE, createRoutingAllocation(state)),
+            Decision.Type.NO,
+            "node [" + DATA_NODE.getId() + "] is preparing to be removed from the cluster"
+        );
+        assertThatDecision(
```

Comment:
Can we add a verification that a shard that is not yet on the source node cannot expand to the target node?

Reply:
Do you mean that shard from any other node of the cluster cannot auto-expand to the replacement target node?
I believe it can according to the `NodeShutdownAllocationDecider`

Reply:
The `NodeReplacementAllocationDecider` should reject this here:

https://github.com/elastic/elasticsearch/pull/96281/files#diff-6c6bb87a8918c913f92d49ba4c9dc5973f35052f062afe45706dc3a0011b324eR153

for the case where the shard is not yet present on the node.

Reply:
I did not find this comment addressed?

Reply:
This scenario:
* put a shard on the node_c
* replace node_a -> node_b
* check shard from the node_c is not auto-expanding replicas to node_a or node_b

is verified in `org.elasticsearch.cluster.routing.allocation.decider.NodeReplacementAllocationDeciderTests#testShouldNotAutoExpandReplicasDuringUnrelatedNodeReplacement`, unless I miss something

Reply:
Thanks for pointing me to it.